packages/fast-components-styles-msft/src/utilities/color/neutral-foreground.ts

```diff
@@ -1,82 +1,130 @@
-            : neutralForegroundDark(designSystem);
-        const restIndex: number = palette.indexOf(restColor);
-        return getSwatch(restIndex + deltaResolver(designSystem) * direction, palette);
-    };
-}
+function neutralForegroundAlgorithm(): DesignSystemResolver<SwatchFamily> {
+    return (designSystem: DesignSystem): SwatchFamily => {
+        const neutralPalette: Palette = palette(PaletteType.neutral)(designSystem);
+        const paletteLength: number = neutralPalette.length;
+        const maxIndex: number = paletteLength - 1;
 
-/**
- * Retrieve light neutral-foreground color for use on dark backgrounds
- */
-export const neutralForegroundLight: SwatchResolver = (
-    designSystem: DesignSystem
-): Swatch => {
-    return getSwatch(
-        neutralForegroundLightIndex(designSystem),
-        neutralPalette(designSystem)
-    );
-};
+        const stateDeltas: any = {
+            rest: 0,
+            hover: neutralForegroundHoverDelta(designSystem),
+            active: neutralForegroundActiveDelta(designSystem),
+        };
 
-/**
- * Retrieve dark neutral-foreground color for use on light backgrounds
- */
-export const neutralForegroundDark: SwatchResolver = (
-    designSystem: DesignSystem
-): Swatch => {
-    return getSwatch(
-        neutralForegroundDarkIndex(designSystem),
-        neutralPalette(designSystem)
-    );
-};
+        const direction: 1 | -1 = isDarkMode(designSystem) ? -1 : 1;
+
+        const accessibleSwatch: Swatch = swatchByContrast(
+            backgroundColor // Compare swatches against the background
+        )(
+            palette(PaletteType.neutral) // Use the neutral palette
+        )(
+            neutralForegroundInitialIndexResolver // Begin searching based on neutral index, direction, and deltas
+        )(
+            () => direction // Search direction based on light/dark mode
+        )(
+            contrastTargetFactory(14) // A swatch is only valid if the contrast is greater than indicated
+        )(
+            designSystem // Pass the design system
+        );
+
+        // One of these will be rest, the other will be hover. Depends on the offsets and the direction.
+        const accessibleIndex1: number = findSwatchIndex(
+            PaletteType.neutral,
+            accessibleSwatch
+        )(designSystem);
+        const accessibleIndex2: number =
+            accessibleIndex1 + direction * Math.abs(stateDeltas.rest - stateDeltas.hover);
+
+        const indexOneIsRestState: boolean =
+            direction === 1
+                ? stateDeltas.rest < stateDeltas.hover
+                : direction * stateDeltas.rest > direction * stateDeltas.hover;
+
+        const restIndex: number = indexOneIsRestState
+            ? accessibleIndex1
+            : accessibleIndex2;
+        const hoverIndex: number = indexOneIsRestState
+            ? accessibleIndex2
+            : accessibleIndex1;
+
+        const restClamped: number = clamp(restIndex, 0, maxIndex);
```

Comment:
This clamping work is redundant and performed by `getSwatch` - you can remove it.